esp-hal/src/gpio/mod.rs

```diff
@@ -2212,34 +2301,80 @@ mod asynch {
     where
         P: InputPin,
     {
-        async fn wait_for(&mut self, event: Event) {
-            self.listen(event);
-            PinFuture::new(self.pin.number()).await
+        /// Wait until the pin experiences a particular [`Event`].
+        ///
+        /// The GPIO driver will disable listening for the event once it occurs,
+        /// or if the `Future` is dropped.
+        ///
+        /// Note that calling this function will overwrite previous
+        /// [`listen`][Self::listen] operations for this pin.
+        #[inline]
+        pub async fn wait_for(&mut self, event: Event) {
+            let mut future = PinFuture {
+                pin: unsafe { self.clone_unchecked() },
+            };
+
+            // Make sure this pin is not being processed by an interrupt handler.
+            if future.pin.is_listening() {
+                future.pin.unlisten();
+                poll_fn(|cx| {
+                    if future.pin.is_interrupt_set() {
+                        cx.waker().wake_by_ref();
+                        Poll::Pending
+                    } else {
+                        Poll::Ready(())
+                    }
+                })
+                .await;
```

Comment:
This can be written as.
```Rust
while future.pin.is_interrupt_set() {
    embassy_futures::yield_now().await;
}
```

Reply:
Yeah, but I'd rather not rely on the optimizer to remove the "has_yielded" flag. We don't need embassy here.